x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    private void addVectors(BytesRefVector v1, BytesRefVector v2, BytesRefVector v3, GroupingAggregatorFunction.AddInput addInput) {
+        final int positionCount = v1.getPositionCount();
+        try (IntVector.Builder ordsBuilder = blockFactory.newIntVectorFixedBuilder(positionCount)) {
+            // TODO: enable ordinal vectors in BytesRefBlockHash
+            try (IntVector k1 = hash1.add(v1); IntVector k2 = hash2.add(v2); IntVector k3 = hash3.add(v3)) {
+                for (int p = 0; p < positionCount; p++) {
+                    long ord = hashOrdToGroup(finalHash.add(k1.getInt(p), k2.getInt(p), k3.getInt(p)));
+                    ordsBuilder.appendInt(Math.toIntExact(ord));
+                }
+            }
+            try (IntVector ords = ordsBuilder.build()) {
+                addInput.add(0, ords);
+            }
+        }
+    }
+
+    private class AddWork extends AbstractAddBlock {
+        final IntBlock b1;
+        final IntBlock b2;
+        final IntBlock b3;
+
+        AddWork(IntBlock b1, IntBlock b2, IntBlock b3, GroupingAggregatorFunction.AddInput addInput) {
+            super(blockFactory, emitBatchSize, addInput);
+            this.b1 = b1;
+            this.b2 = b2;
+            this.b3 = b3;
+        }
+
+        void add() {
+            int positionCount = b1.getPositionCount();
+            for (int i = 0; i < positionCount; i++) {
+                int v1 = b1.getValueCount(i);
+                int v2 = b2.getValueCount(i);
+                int v3 = b3.getValueCount(i);
+                int first1 = b1.getFirstValueIndex(i);
+                int first2 = b2.getFirstValueIndex(i);
+                int first3 = b3.getFirstValueIndex(i);
+                if (v1 == 1 && v2 == 1 && v3 == 1) {
+                    long ord = hashOrdToGroup(finalHash.add(b1.getInt(first1), b2.getInt(first2), b3.getInt(first3)));
+                    ords.appendInt(Math.toIntExact(ord));
+                    addedValue(i);
+                    continue;
+                }
+                ords.beginPositionEntry();
+                for (int i1 = 0; i1 < v1; i1++) {
+                    int k1 = b1.getInt(first1 + i1);
+                    for (int i2 = 0; i2 < v2; i2++) {
+                        int k2 = b2.getInt(first2 + i2);
+                        for (int i3 = 0; i3 < v3; i3++) {
+                            int k3 = b3.getInt(first3 + i3);
+                            long ord = hashOrdToGroup(finalHash.add(k1, k2, k3));
+                            ords.appendInt(Math.toIntExact(ord));
+                            addedValueInMultivaluePosition(i);
+                        }
+                    }
+                }
+                ords.endPositionEntry();
+            }
+            emitOrds();
+        }
+    }
+
+    @Override
+    public ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize) {
+        throw new UnsupportedOperationException("TODO");
```

Comment:
Sorry @nik9000. I will implement this in a follow-up.

Reply:
None of the specialize versions have an implementation of it at the moment. It's kind of a pain to implement. I've got to make some building blocks to help it.